### DISCUSSION TITLE
REQUEST: Toggle mount power on/off using relay via MGPBox?

### DISCUSSION THREAD

**Discussion by airscottdenning** [mworion/MountWizzard4#151] [Feature Request]
The purpose of the "P" in MGPBox is to send a 1500 ms pulse over a DC cable to the relay in the 10Micron control box. TheMGPBox is serial-connected to the mount for GPS and can be separately connected via FTDI serial on USB to the computer.

In DOS/Windows, the command "MGBox.exe -P COM?? -T 1500" causes the box to send the required pulse to the relay on the mount control box, safely toggling power on or off. Here COM?? is the Windows COM port assigned to the MGPBox (actually an FTDI usb-to-serial converter).

I have tried this in wine, but I can't figure out how to map the /dev/ttyUSBx or /dev/ttySx port in linux to the MGPBox.

This seems like a nice piece of functionality for MW4.

Sorry, I didn't know whether t post this under Discussions or Issues, so I did both.

**Comment by mworion** [comment]
Hi Scott,

actually MW4 uses WOL for boot and the 10micron command protocol for shutdown. If you need a "hard" version of switch, MW4 already supports KMRelay IP relain box and you could configure the outputs "Pulsed" as well. I did use this for switch on off, too.
Right now I have not seen (as I own only a MGBox) if there is a ASCOM Switch driver for MGPBox, which would enable me adding the box as On / Off switch as well. MW4 as it is based on python is not designed to emulate low level drivers as there might be runtime or timing constraints. 

Sorry so far
Michel

**Comment by airscottdenning** [comment]
Hi Michel,

I was able to get this to work. Appended below is a very short python script that toggles power to the mount using the USB serial port in the MGPBox.

Essentially I had to poke around till I found the device in /dev and then just send the command ':pulse,1500*\n’ over its serial port.

I have used the WOL functionality in MW4, but it only works for me if there is already power to the mount. The python script below allows me to power up from completely off once I’ve triggered the WiFi outlet that powers my equipment. This was the last step for me. Since my scope is in the back garden it’s easy enough to just walk out there and rock the physical switch on the controller, but in theory this would work across the world.

The MGPBox provides GPS and meteo data to the mount via its own serial cable (just as the MGBox does), but adds a simple DC cable to the relay on the 10Micron controller.

Thanks again for your amazing work on MW4!

Scott

=======  BEGIN SCRIPT

#!/usr/bin/env python3
# This simple script opens a serial port to the MGPBox and sends a 1.5 second pulsed DC
# signal to toggle power on or off for the 10Micron mount to which the MGPBox is connected

import serial

# Set the device for the USB serial port on the MGPBox
MGPBox = '/dev/FTDIusb-D307KUXX'

# This is the serial command to toggle 10Micron power
pulseCommand = ':pulse,1500*\n'

# Open port
serialCon = serial.Serial(port=MGPBox, baudrate=38400)

# Send pulse to relay
serialCon.write(str.encode(pulseCommand))

# Close the port
serialCon.close()

**Reply by mbalkham** [reply]
Hi Scott, any idea how to run this script (or similar) within a voyager dragscript (go easy on me if really straight forward, scripting is all new to me!)

**Comment by airscottdenning** [comment]
I’m just starting to learn voyager so haven’t run scripts in there yet. 

I do indeed have the 10Micron “toggle power” script working in windows, but I had to make a change to get it to talk to the serial port (COM4 in my case). I’ve been just double clicking the script icon on the desktop and that seems to work great for powering the 10Micron on and off.

I can send the revised script to you, but you’ll need to make a small mod unless your powerboat happens to be on COM4. I still have to learn how to connect it to voyager.

**Reply by RawNoise** [reply]
Hi Scott, I would love to try out your script on my 10u. I'm running my rig at a remote observatory and use a MGPBox for power but a one click solution  would solve an automation problem I have. Scripting is new to me but would love to give it a try. Thanks in advance!!

Steve - Raw Noise Astro

**Comment by mbalkham** [comment]
I'm running this in a voyager dragscript using viking and dragonfly to pulse the mount to power on and power off (rather than through the MGP box).

**Reply by RawNoise** [reply]
Gotcha! I'm looking to run something like this in NINA. The extra steps with the MGPBox hinder my ability to fully automate my sequences. I really wish 10u or MW4 could/would implement a better solution.

**Comment by mbalkham** [comment]
Depending on your definition of 'fully automate' you might need something other than NINA anyway. We run fully autonomosly under a perpetual script needing no input (other than a list of targets and expsosures). If you are talking 'just' about remotely controlling on a per night basis or automating locally than NINA will prob be fine. Also NINA is constantly adding features and refining so might be posible these days to go the fully hog. You might try on the 10m forum. Sure there will be other NINA users that can offer advice.

**Comment by airscottdenning** [comment]
Attached below is the text of my python script to power the 10u mount on or off (or just toggle power). I've also attached to this message as a plain text file. You need to be careful to get the indentation of each line exactly right.

I believe everything in here is standard python3.\

You need to set two parameters at the head of the file that are unique to your system:
1) set "mountIP" to the IP address of your mount on your local network
2) set "MGPBox" to the serial port of your MGPBox (on Windows it's a COM port and on linux it's a file in /dev)

The script needs the IP address of the mount to determine whether it's currently on or off
The script needs the serial port to send the pulse that toggles power

If you're on Windows, save the script someplace like "C:\USERS\yourname\10u.py"

To power the mount on, cd to the directory in which the script is stored (or include this in your PATH), 
then at the Windows command line, type "10u ON" or "10u on"

Similarly, to power the mount off, type "10u off"

I've created a shortcut on my Windows Desktop that just runs the script without any arguments. To be fancy I gave it the 10u icon so I remember what it does. 

If the mount is powered off, then when I click the shortcut icon, the mount powers on.
If the mount is powered on, then when I click the shortcut icon, the mount powers off.

In NINA or in a .bat script, turn on the power with the command:
c:\Users\yourname\10u.power.py ON

Please feel free to message me anytime for help.
Or email airscottdenning@yahoo.com

[10u.py.txt](https://github.com/user-attachments/files/18399134/10u.py.txt)

==========================================================================
```
#!/usr/bin/env python3

# This script is used to power the 10Micron mount on or off
# Usage "10u on" powers on the mount ... any other argument toggles power

import time
import sys

# Set IP address of the mount here
mountIP = "192.168.10.145"

# Set the device for the USB serial port on the MGPBox
#MGPBox = '/dev/FTDIusb-D307KUXX'
MGPBox = 'COM3'

##################

# Function uses the MGPBox to toggle power on 10Micron mount

def toggle(): 

	import serial
	import time

	# This is the serial command to toggle 10Micron power
	pulseCommand = ':pulse,1500*\n'

	# Open port
	serialCon = serial.Serial(port=MGPBox, baudrate=38400)

	# Send pulse to relay
	time.sleep(2)
	serialCon.write(str.encode(pulseCommand))
	time.sleep(2)

	# Tell the user that we're done
	print('Toggled power to 10Micron Mount')

	# Close the port
	serialCon.close()

##################

# Function uses the ping to test for ethernet connectivity of mount

def mountIsAwake(mountIP):

	# Use ping to see if the mount is awake
	# Return code is zero if mount is awake, nonzero if no connection

	import subprocess

	output = subprocess.Popen(["ping.exe","-n", "1",mountIP],stdout = subprocess.PIPE).communicate()[0]

	if (b'TTL' in output):
		return(1)
	else:
		return(0)
	
##################

# Main program

if len(sys.argv) == 2: # If there's a command-line argument, figure out what to do

	# Check command line argument
	if sys.argv[1] == 'on' or sys.argv[1] == 'ON':

		# User has requested that the mount be POWERED ON

		# First make sure mount isn't already on & awake
		if mountIsAwake(mountIP):
			print("Mount is already awake")

		else:

			# Acknowledge useer request
			print("Powering on the mount")

			# Toggle power to on
			toggle()

			# Give the mount 40 seconds to boot up
			print("Waiting 40 seconds for mount to wake up ...")
			time.sleep(40)

			# Make sure it worked & report to user
			if mountIsAwake(mountIP):
				print("Mount is on and connected at ", mountIP)
			else:
				print("Mount is not connected. Try again in a few seconds.")

	elif sys.argv[1] == 'off' or sys.argv[1] == 'OFF':	

		# User has requested that the mount be POWERED OFF

		# If mount is on & awake, turn it off
		if mountIsAwake(mountIP):
			toggle()
		else: 
			print("Mount is already off")

	else:

		# Any other argument triggers help message
		print("Usage 10u [ON | OFF]")

else: # If there's no command-line argument, just toggle the power

	toggle()
```

**Comment by airscottdenning** [comment]
Attached below is the text of my python script to power the 10u mount on or off (or just toggle power). I've also attached to this message as a plain text file. You need to be careful to get the indentation of each line exactly right.

I believe everything in here is standard python3.

You need to set two parameters at the head of the file that are unique to your system:
set "mountIP" to the IP address of your mount on your local network
set "MGPBox" to the serial port of your MGPBox (on Windows it's a COM port and on linux it's a file in /dev)

The script needs the IP address of the mount to determine whether it's currently on or off.

The script needs the serial port to send the pulse that toggles power.

If you're on Windows, save the script someplace like "C:\USERS\yourname\10u.py”

To power the mount on, cd to the directory in which the script is stored (or include this in your PATH),
then at the Windows command line, type "10u ON" or "10u on”

Similarly, to power the mount off, type "10u off”

I've created a shortcut on my Windows Desktop that just runs the script without any arguments. To be fancy I gave it the 10u icon so I remember what it does.

If the mount is powered off, then when I click the shortcut icon, the mount powers on.
If the mount is powered on, then when I click the shortcut icon, the mount powers off.

In NINA or in a .bat script, turn on the power with the command:
c:\Users\yourname\10u.power.py ON

Please feel free to message me anytime for help.
Or reply to this email ***@***.***)

##############################################################

#!/usr/bin/env python3

# This script is used to power the 10Micron mount on or off
# Usage "10u on" powers on the mount ... any other argument toggles power

import time
import sys

# Set IP address of the mount here
mountIP = "192.168.10.145"

# Set the device for the USB serial port on the MGPBox
#MGPBox = '/dev/FTDIusb-D307KUXX'
MGPBox = 'COM3'

##################

# Function uses the MGPBox to toggle power on 10Micron mount

def toggle(): 

	import serial
	import time

	# This is the serial command to toggle 10Micron power
	pulseCommand = ':pulse,1500*\n'

	# Open port
	serialCon = serial.Serial(port=MGPBox, baudrate=38400)

	# Send pulse to relay
	time.sleep(2)
	serialCon.write(str.encode(pulseCommand))
	time.sleep(2)

	# Tell the user that we're done
	print('Toggled power to 10Micron Mount')

	# Close the port
	serialCon.close()

##################

# Function uses the ping to test for ethernet connectivity of mount

def mountIsAwake(mountIP):

	# Use ping to see if the mount is awake
	# Return code is zero if mount is awake, nonzero if no connection

	import subprocess

	output = subprocess.Popen(["ping.exe","-n", "1",mountIP],stdout = subprocess.PIPE).communicate()[0]

	if (b'TTL' in output):
		return(1)
	else:
		return(0)
	
##################

# Main program

if len(sys.argv) == 2: # If there's a command-line argument, figure out what to do

	# Check command line argument
	if sys.argv[1] == 'on' or sys.argv[1] == 'ON':

		# User has requested that the mount be POWERED ON

		# First make sure mount isn't already on & awake
		if mountIsAwake(mountIP):
			print("Mount is already awake")

		else:

			# Acknowledge useer request
			print("Powering on the mount")

			# Toggle power to on
			toggle()

			# Give the mount 40 seconds to boot up
			print("Waiting 40 seconds for mount to wake up ...")
			time.sleep(40)

			# Make sure it worked & report to user
			if mountIsAwake(mountIP):
				print("Mount is on and connected at ", mountIP)
			else:
				print("Mount is not connected. Try again in a few seconds.")

	elif sys.argv[1] == 'off' or sys.argv[1] == 'OFF':	

		# User has requested that the mount be POWERED OFF

		# If mount is on & awake, turn it off
		if mountIsAwake(mountIP):
			toggle()
		else: 
			print("Mount is already off")

	else:

		# Any other argument triggers help message
		print("Usage 10u [ON | OFF]")

else: # If there's no command-line argument, just toggle the power

	toggle()

￼

> On Jan 12, 2025, at 9:11 PM, RawNoise ***@***.***> wrote:
> 
> 
> Hi Scott, I would love to try out your script on my 10u. I'm running my rig at a remote observatory and use a MGPBox for power but a one click solution would solve an automation problem I have. Scripting is new to me but would love to give it a try. Thanks in advance!!
> 
> Steve - Raw Noise Astro
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/mworion/MountWizzard4/discussions/151#discussioncomment-11815936>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABF6OHZQ3TMZPA7E65PFAPD2KM4HJAVCNFSM6AAAAABVBX3YKSVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCOBRGU4TGNQ>.
> You are receiving this because you authored the thread.
> 


#!/usr/bin/env python3

# This script is used to power the 10Micron mount on or off
# Usage "10u on" powers on the mount ... any other argument toggles power

import time
import sys

# Set IP address of the mount here
mountIP = "192.168.10.145"

# Set the device for the USB serial port on the MGPBox
#MGPBox = '/dev/FTDIusb-D307KUXX'
MGPBox = 'COM3'

##################

# Function uses the MGPBox to toggle power on 10Micron mount

def toggle(): 

	import serial
	import time

	# This is the serial command to toggle 10Micron power
	pulseCommand = ':pulse,1500*\n'

	# Open port
	serialCon = serial.Serial(port=MGPBox, baudrate=38400)

	# Send pulse to relay
	time.sleep(2)
	serialCon.write(str.encode(pulseCommand))
	time.sleep(2)

	# Tell the user that we're done
	print('Toggled power to 10Micron Mount')

	# Close the port
	serialCon.close()

##################

# Function uses the ping to test for ethernet connectivity of mount

def mountIsAwake(mountIP):

	# Use ping to see if the mount is awake
	# Return code is zero if mount is awake, nonzero if no connection

	import subprocess

	output = subprocess.Popen(["ping.exe","-n", "1",mountIP],stdout = subprocess.PIPE).communicate()[0]

	if (b'TTL' in output):
		return(1)
	else:
		return(0)
	
##################

# Main program

if len(sys.argv) == 2: # If there's a command-line argument, figure out what to do

	# Check command line argument
	if sys.argv[1] == 'on' or sys.argv[1] == 'ON':

		# User has requested that the mount be POWERED ON

		# First make sure mount isn't already on & awake
		if mountIsAwake(mountIP):
			print("Mount is already awake")

		else:

			# Acknowledge useer request
			print("Powering on the mount")

			# Toggle power to on
			toggle()

			# Give the mount 40 seconds to boot up
			print("Waiting 40 seconds for mount to wake up ...")
			time.sleep(40)

			# Make sure it worked & report to user
			if mountIsAwake(mountIP):
				print("Mount is on and connected at ", mountIP)
			else:
				print("Mount is not connected. Try again in a few seconds.")

	elif sys.argv[1] == 'off' or sys.argv[1] == 'OFF':	

		# User has requested that the mount be POWERED OFF

		# If mount is on & awake, turn it off
		if mountIsAwake(mountIP):
			toggle()
		else: 
			print("Mount is already off")

	else:

		# Any other argument triggers help message
		print("Usage 10u [ON | OFF]")

else: # If there's no command-line argument, just toggle the power

	toggle()

**Reply by RawNoise** [reply]
Scott, 
Thanks so much for sharing your script with me. This is the last piece of my fully remote operation puzzle. 

I have to admit I'm not a python wizard and I'm getting this error after creating and running the script......

C:\Users\PrimaLuceLab\Documents\Python Scripts>10u.power.py
Traceback (most recent call last):
  File "C:\Users\PrimaLuceLab\Documents\Python Scripts\10u.power.py", line 110, in <module>
    toggle()
  File "C:\Users\PrimaLuceLab\Documents\Python Scripts\10u.power.py", line 22, in toggle
    import serial
ModuleNotFoundError: No module named 'serial'

**Comment by airscottdenning** [comment]
Looks like I was wrong about everything being “standard python” and you will need to install the “pyserial” package in your python.

At the Windows command line try typing
“pip install pyserial”

Pip is a python package manager. If pip isn’t installed on your system you may need to install pip first. Here’s a howto article:

https://www.geeksforgeeks.org/how-to-install-pip-on-windows/

Please feel free to email me directly ***@***.***

**Reply by RawNoise** [reply]
Well that did it, Scott! I now have a functional python script toggling the power on and off. Thank you so much for your help. I really appreciate it and this will do wonders for my remote observatory. I'm grateful for your time and patience with me. Thanks again.

**Comment by airscottdenning** [comment]
Hooray! Glad to be able to help.

Best wishes,
Scott

**Comment by gnthibault** [comment]
Thank you so much @airscottdenning I am not a windows user, but that's really handy, as I am building fully autonomous python-controlled observatory.
